crates/compiler/src/visitors/type_check_visitor.rs

```diff
@@ -0,0 +1,945 @@
+    format!("Can't figure out the type of variable {name} given its context. Specify its type with a <<declare>> statement.")
+}
+
+fn get_filename(path: &str) -> &str {
+    if let Some(os_str) = Path::new(path).file_name() {
+        if let Some(file_name) = os_str.to_str() {
+            return file_name;
+        }
+    }
+    path
+}
+
+trait GetHashableInterval<'input>: ParserRuleContext<'input> {
+    fn get_hashable_interval(&self) -> HashableInterval {
+        let interval = self.get_source_interval();
+        HashableInterval(interval)
+    }
+}
+
+impl<'input, T: ParserRuleContext<'input>> GetHashableInterval<'input> for T {}
+
+trait DefaultValue {
+    fn default_value(&self) -> Option<Convertible>;
+}
+impl DefaultValue for Type {
+    fn default_value(&self) -> Option<Convertible> {
+        match self {
+            Type::String => Some(Convertible::String(Default::default())),
+            Type::Number => Some(Convertible::Number(Default::default())),
+            Type::Boolean => Some(Convertible::Boolean(Default::default())),
+            _ => None,
+        }
+    }
+}
+
+impl DefaultValue for Option<Type> {
+    fn default_value(&self) -> Option<Convertible> {
+        self.as_ref()?.default_value()
+    }
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+struct HashableInterval(Interval);
+
+impl From<Interval> for HashableInterval {
+    fn from(interval: Interval) -> Self {
+        Self(interval)
+    }
+}
+
+impl Ord for HashableInterval {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.0.a.cmp(&other.0.a).then(self.0.b.cmp(&other.0.b))
+    }
+}
+
+impl PartialOrd for HashableInterval {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Hash for HashableInterval {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.a.hash(state);
+        self.0.b.hash(state);
+    }
+}
+
+impl Deref for HashableInterval {
+    type Target = Interval;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for HashableInterval {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn allows_valid_assignments() {
+        let file = File {
+            file_name: "test.yarn".to_string(),
+            source: "title: test
+---
+<<declare $foo to 1>>
+<<set $foo to 2>>
+<<declare $bar to true>>
+<<declare $baz to \"hello\">>
+<<set $bar to false>>
+<<set $baz to \"world\">>
+==="
+            .to_string(),
+        };
+        let result = compile(CompilationJob {
+            files: vec![file],
+            library: None,
+            compilation_type: CompilationType::FullCompilation,
+            variable_declarations: vec![],
+        });
+
+        println!("{:?}", result.diagnostics);
+        assert!(result.diagnostics.is_empty());
+    }
+
+    #[test]
```

Comment:
Can we put the tests into a `tests` module so that they're not compiled in production?

Reply:
Check lines 719 and 720, they already are :)

Reply:
Oh whoops, I'm not sure how I missed that :sweat_smile: